File: shared/types/optimization-run.ts
import { Timestamp } from 'firebase-admin/firestore';

/** 最適化実行ステータス */
export type OptimizationStatus = 'Optimal' | 'Feasible' | 'Infeasible' | 'Not Solved';

/** 割当結果（1オーダー） */
export interface AssignmentRecord {
  order_id: string;
  staff_ids: string[];
}

/** 最適化実行パラメータ */
export interface OptimizationParameters {
  time_limit_seconds: number;
}

/** 最適化実行記録 */
export interface OptimizationRun {
  id: string;
  week_start_date: Timestamp;
  executed_at: Timestamp;
  executed_by: string;
  dry_run: boolean;
  status: OptimizationStatus;
  objective_value: number;
  solve_time_seconds: number;
  total_orders: number;
  assigned_count: number;
  assignments: AssignmentRecord[];
  parameters: OptimizationParameters;
}

File: optimizer/src/optimizer/models/optimization_run.py
"""最適化実行記録モデル"""

from datetime import datetime

from pydantic import BaseModel

from .problem import Assignment


class OptimizationParameters(BaseModel):
    """最適化パラメータ"""

    time_limit_seconds: int
    w_travel: float = 1.0
    w_preferred_staff: float = 5.0
    w_workload_balance: float = 10.0
    w_continuity: float = 3.0


class OptimizationRunRecord(BaseModel):
    """Firestoreに保存する最適化実行記録"""

    id: str
    week_start_date: str  # YYYY-MM-DD
    executed_at: datetime
    executed_by: str  # Firebase Auth UID
    dry_run: bool
    status: str  # "Optimal", "Feasible", "Infeasible", "Not Solved"
    objective_value: float
    solve_time_seconds: float
    total_orders: int
    assigned_count: int
    assignments: list[Assignment]
    parameters: OptimizationParameters
